### ISSUE TITLE
Icon for cell move handle

### ISSUE THREAD

**Issue by jweill-aws** [jupyterlab/jupyterlab#12188]
@aiqc observed in https://github.com/jupyterlab/jupyterlab/pull/12028#issuecomment-1062278035 that "there is already functionality in the UI to move cells around via the mouse, but it just isn't apparent. When you hover over the little landing zone to the left side of the cell, a tiny 4-way arrow appears. It lets you click and drag the cell up/ down to reposition it."

### Problem

The "move cell" handle element is active, but invisible. The only way to know that you've moused over it is that the mouse cursor changes to a four-way arrow. (Image courtesy @aiqc)

![A mouse cursor over the invisible resize handle on a cell](https://user-images.githubusercontent.com/74990642/157335181-e15990ac-90f8-4c0e-9ec8-048fa22f4503.png)

### Proposed Solution

Add an icon to indicate that the user can click/tap and drag up and down to reposition this cell. (Icon samples courtesy @aiqc)

![Sample icons to put in the move handle area](https://user-images.githubusercontent.com/74990642/157335198-7e9a54b8-956e-490a-8155-1feeed6c3a23.png)

### Additional context

Open questions:
* Should the icon be visible when the current cell is active? Should it appear on hover?
* Should the icon be visible on inactive cells when the user hovers over the cell move handle area?

**Comment by krassowski** [comment]
Question: where would you propose this icon to appear?

I would say that the space is rather limited and personally I would prefer not to have such an icon, or to have an option to disable it. As described above, there already is a cursor with the move (4-way arrow) icon, so once user hovers over this area they will know that they can move the cell. Further, the action to move cells is already well documented ([here](https://jupyterlab.readthedocs.io/en/stable/user/notebook.html)), so a user who needs it will read the documentation and see the following animation: https://www.youtube.com/watch?v=J9xoTGdqWIo. I would be more inclined if this proposal was based on user feedback but it appears to be a side discussion when implementing a new functionality, so maybe it is something which is not crucial?

However, the dedicated move icon would a good idea for the mobile layout:
- users on mobile devices with touchscreens do not see the cursor move icon (there is no mouse!)
- since we already move the execution counter to a separate row we will have plenty of space to add the icon

Yet if https://github.com/jupyterlab/jupyterlab/pull/12028#issuecomment-1063312331 keeps the up/down arrows in the toolbar then we don't need it for mobile layout either.

**Comment by aiqc** [comment]
The functionality is too difficult to discover in comparison to how important it is. That's why people were trying to give it a higher priority in that new bar.

"when you hover over this small, invisible area, that you would otherwise have no reason to go to - you get a hint that it does something.

no one reads that documentation

**Comment by krassowski** [comment]
> Add an icon to indicate that the user can click/tap and drag up and down to reposition this cell. 
> ![image](https://user-images.githubusercontent.com/5832902/157698142-9a05d42b-70f9-4537-b35e-87c5f467e0a7.png)

- these icons suggest that only up/down motion is possible, whereas this functionality also allows to move cells between notebooks too (currently making a copy). The four-way arrow seems more appropriate for that reason.
- multi-select: would only one icon be shown when selecting multiple cells? If it was shown on hover, would it show over the hovered cell or over all selected cells over which the cursor hovered?

This is the current behaviour:

![move-multiple](https://user-images.githubusercontent.com/5832902/157699190-c713c2b6-8664-4515-8672-c8f36e55e553.gif)

**Comment by jweill-aws** [comment]
Thanks all for the feedback — @krassowski I had no idea that this control let you copy cells across notebooks!

I still think that there is value in a visible "handle" in a cell to indicate where I can click to move things around. The title bars on desktop OSes (ambiguous as they are) are defined, bounded zones where I am confident that I can click and drag to move the window around. I expect resize handles to appear when I mouse at the edges of a non-full-screened window. The boundary of a cell, by design, does not have a visible zone where I would expect the cursor to work as a move cursor. I still find myself clicking on the blue "collapse/expand" handle to try moving a cell around, to no avail.

**Comment by jweill-aws** [comment]
Notes from triage:
* Consider adding label (tooltip) when the cursor turns into a 4-way "move" cursor
* Consider having an outline visible, perhaps on hover, to indicate that the user is hovering over a control that affects the cell
* Mobile implications? (Too many icons on mobile will be a distracting use of space)

**Comment by krassowski** [comment]
Hijacking a screenshot of interactive design created for unrelated purposes in https://github.com/jupyterlab/jupyterlab/issues/9733#issuecomment-1028275057:

![Screenshot from 2022-03-15 22-43-43](https://user-images.githubusercontent.com/5832902/158484743-c846338c-1558-49de-bf6d-1f39f3e39c31.png)

What if the three dots in the middle (which would show on hover over cell) would serve as a handle for dragging the cell around? This would conflict with the proposed placement of cell toolbar in the middle of the cell though.

**Comment by aiqc** [comment]
> three dots in the middle (which would show on hover over cell) would serve as a handle

I like the 3 textured dot concept. If I google "ui component rearrange" it shows a few examples with that dots and hamburgers. However, the grabber feels out of place in the middle. Most of the rearrange components are at the left/right (start/end) of the row. 

Could the blue "selected cell" indicator get those dots or something to show it is grabby?


p.s. Can we get rid of the `:` ? use an underlined `#` instead of `[#]`? or is that coming from ipython

**Comment by jweill-aws** [comment]
For many years, iOS has used the three horizontal lines icon commonly known as a "hamburger menu" to indicate that the user can tap and drag an item to move it. As an example, see this screen shot from the default iOS Mail app in "Edit" mode. (Sorry about inadvertently encoding it as a JPEG.)

![File](https://user-images.githubusercontent.com/93281816/158652926-44a9bb52-d6d6-4614-8d10-14b37c84eacb.jpg)

**Comment by krassowski** [comment]
> Most of the rearrange components are at the left/right (start/end) of the row.

Yes, I know. Thinking more about it now, if the visual clue was on the right border of the cell, and only shown on hover it could take very little extra space (say 2-3 pixels and the rest would just overlay the current padding). The right side does not have any existing functions (such as collapsing cell, or output, as there are on the left side) so I would be happy with such a solution in general.

> Could the blue "selected cell" indicator get those dots or something to show it is grabby? hmm. it runs the length of the cell. so would you center the dots or place them at the top. maybe not.

This is a functional component which collapses the cell on click so it would be misleading to have a grabber icon on it.

> p.s. Can we get rid of the []: characters ? use an underlined # instead of [#]? guessing it comes from ipython cli where cells are less apparent and the []: is more needed

This is a nice thought, though I am not sure if we are ready for such a revolution. I guess we could try to start gradually by dropping `:` ;)
